### PR TITLE
Implement `status.uptime` on macOS

### PR DESCRIPTION
**Pull request by jfindlay**
### What does this PR do?

Implement `status.uptime` on macOS
### What issues does this PR fix or reference?
#37158
### Tests written?

Yes

### CODE REVIEW

**Comment by cachedout** [comment]
Go Go Jenkins!

File: salt/modules/status.py [hunk 1]
```diff
@@ -148,26 +148,44 @@ def uptime():
...
     if salt.utils.is_linux():
         ut_path = "/proc/uptime"
         if not os.path.exists(ut_path):
             raise CommandExecutionError("File {ut_path} was not found.".format(ut_path=ut_path))
-        ut_ret['seconds'] = int(float(open(ut_path).read().strip().split()[0]))
+        seconds = float(salt.utils.fopen(ut_path).read().split()[0])
     elif salt.utils.is_sunos():
         cmd = "kstat -p unix:0:system_misc:boot_time | nawk '{printf \"%d\\n\", srand()-$2}'"
-        ut_ret['seconds'] = int(__salt__['cmd.shell'](cmd, output_loglevel='trace').strip() or 0)
+        seconds = float(__salt__['cmd.shell'](cmd, output_loglevel='trace').strip() or 0)
```
Comment (isbm) [inline salt/modules/status.py]: On Solaris, with `kstat` you will always get an integer, no need to `float` here.

File: salt/modules/status.py [hunk 2]
```diff
@@ -148,26 +148,44 @@ def uptime():
...
+    elif salt.utils.is_darwin():
+        bt_data = __salt__['sysctl.get']('kern.boottime')
+        if not bt_data:
+            raise CommandExecutionError('Cannot find kern.boottime system parameter')
+        sec_data, usec_data = bt_data.split('}')[0].strip(' {').split(', ')
+        sec = sec_data.split('sec = ')[1]
+        usec = usec_data.split('usec = ')[1]
```
Comment (isbm) [inline salt/modules/status.py]: I would not do that and suggest something more reliable, although probably less readable:
```python
data = bt_data.split("{")[-1].split("}")[0].strip().replace(' ', '')
uptime = dict([(k_sec, int(v_sec,))
                for k_sec, v_sec in [pair.strip().split("=")
                                     for pair in data.split(",")]])
```
What you do here is to turn mac returned data into a dictionary, and refer to it something like:
```python
uptime['usec']
uptime['sec']
```

You can turn the above to a lambda and also query sleep time etc.

File: salt/modules/status.py [hunk 3]
```diff
@@ -148,26 +148,44 @@ def uptime():
...
+        seconds = curr_seconds - float(sec + '.' + '{0:0>6}'.format(usec))
```
Comment (isbm) [inline salt/modules/status.py]: BTW, Mac is also only an integer, no floats there. :smiley:

**Review by isbm — CHANGES_REQUESTED** [review]
@jfindlay please look at these.

**Comment by isbm** [comment]
@cachedout please hold a moment. :smile:

**Comment by cachedout** [comment]
@isbm Will do. :]

**Comment by isbm** [comment]
@jfindlay two more to go. :-)

**Comment by cachedout** [comment]
@jfindlay Could you please let me know where you are with this one?

**Comment by cachedout** [comment]
Because there has been no response, I am closing this. @jfindlay when you are ready to go here please re-open it.

File: salt/modules/status.py [hunk 4]
```diff
@@ -148,26 +148,42 @@ def uptime():
...
     elif salt.utils.is_sunos():
         cmd = "kstat -p unix:0:system_misc:boot_time | nawk '{printf \"%d\\n\", srand()-$2}'"
-        ut_ret['seconds'] = int(__salt__['cmd.shell'](cmd, output_loglevel='trace').strip() or 0)
+        seconds = int(float(__salt__['cmd.shell'](cmd, output_loglevel='trace').strip() or 0))
+    elif salt.utils.is_darwin():
+        bt_data = __salt__['sysctl.get']('kern.boottime')
+        if not bt_data:
+            raise CommandExecutionError('Cannot find kern.boottime system parameter')
+        sec_data, usec_data = bt_data.split('}')[0].strip(' {').split(', ')
+        seconds = int(curr_seconds - int(sec_data.split('sec = ')[1]))
```
Comment (isbm) [inline salt/modules/status.py]: @jfindlay If you could use my previous suggestion, we by now would also have sleep time of Mac!

**Review by isbm — CHANGES_REQUESTED** [review]
@jfindlay Thanks for the update! Could you please still look at this one?

File: salt/modules/status.py (hunk 4 above)
Comment (isbm) [inline salt/modules/status.py]: Specifically:
```python
data = bt_data.split("{")[-1].split("}")[0].strip().replace(' ', '')
uptime = dict([(k_sec, int(v_sec,))
                for k_sec, v_sec in [pair.strip().split("=")
                                     for pair in data.split(",")]])
```
What you do here is to turn mac returned data into a dictionary, and refer to it something like:
```python
uptime['usec']
uptime['sec']
```

You can turn the above to a lambda and also query sleep time etc.

Comment (isbm) [inline salt/modules/status.py]: Point is, you could reuse it and get also other values from the `sysctl`, such as "sleeptime" or "waketime". That said, you could turn the code above to a `def _internal_function(data)` and reuse it.

Comment (jfindlay) [inline salt/modules/status.py]: @isbm, you're welcome to open a pull request on my branch, `https://github.com/jfindlay/salt/compare/mac_time...isbm:mac_time`.

Comment (isbm) [inline salt/modules/status.py]: Ah, but that's easy:
```python
def _parse_sysctl(data):
    data = data.split("{")[-1].split("}")[0].strip().replace(' ', '')
    return dict([(k_sec, int(v_sec,))
                    for k_sec, v_sec in [pair.strip().split("=")
                                         for pair in data.split(",")]])

...
seconds = int(curr_seconds - _parse_sysctl(bt_data)['sec'])
```

At the moment I don't have Salt installed on OSX.

**Comment by isbm** [comment]
@jfindlay @cachedout I find https://github.com/saltstack/salt/pull/37321 superseding this one. Maybe it would be possible to collaborate with @sjorge ? He did really awesome job.

**Comment by sjorge** [comment]
@isbm ooh I did not know there was a sysctl module, merging stuff now, OSX == FreeBSD it seems.

So we have 4 big buckets: Linux, SunOS, FreeBSD+OSX, OpenBSD+NetBSD... let me try and fold @jfindlay stuff inside mine

**Comment by cachedout** [comment]
I'm fine with using the PR from @sjorge if that's what's agreed on. Any objection, @jfindlay ?

**Comment by jfindlay** [comment]
@cachedout, none from me. Excellent job everyone.

**Pull request closed without merging**